agent/consul/state/catalog.go

```diff
@@ -2546,125 +2546,109 @@ func (s *Store) updateTerminatingGatewayNamespace(tx *memdb.Txn, gateway structs
 	// Check if mapping already exists in table if it's already in the table
 	// Avoid insert if nothing changed
-	existing, err := s.serviceTerminatingGateway(tx, service, entMeta)
+	existing, err := tx.First(gatewayServicesTableName, "id", mapping.Gateway, mapping.Service)
 	if err != nil {
 		return fmt.Errorf("gateway service lookup failed: %s", err)
 	}
 	if gs, ok := existing.(*structs.GatewayService); ok && gs != nil {
+		mapping.CreateIndex = gs.CreateIndex
 		if gs.IsSame(mapping) {
 			return nil
 		}
+	} else {
+		// We have a new mapping
+		mapping.CreateIndex = idx
 	}
+	mapping.ModifyIndex = idx
 
-	if err := tx.Insert(terminatingGatewayServicesTableName, mapping); err != nil {
+	if err := tx.Insert(gatewayServicesTableName, mapping); err != nil {
 		return fmt.Errorf("failed inserting gateway service mapping: %s", err)
 	}
 
-	if err := indexUpdateMaxTxn(tx, idx, terminatingGatewayServicesTableName); err != nil {
-		return fmt.Errorf("failed updating terminating-gateway-services index: %v", err)
+	if err := indexUpdateMaxTxn(tx, idx, gatewayServicesTableName); err != nil {
+		return fmt.Errorf("failed updating gateway-services index: %v", err)
 	}
 	return nil
 }
 
-func (s *Store) serviceTerminatingGateway(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (interface{}, error) {
-	return tx.First(terminatingGatewayServicesTableName, "service", structs.NewServiceID(name, entMeta))
+// serviceGateways returns all GatewayService entries with the given service name. This effectively looks up
+// all the gateways mapped to this service.
+func (s *Store) serviceGateways(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get(gatewayServicesTableName, "service", structs.NewServiceID(name, entMeta))
 }
 
-func (s *Store) terminatingGatewayServices(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
-	return tx.Get(terminatingGatewayServicesTableName, "gateway", structs.NewServiceID(name, entMeta))
+func (s *Store) gatewayServices(tx *memdb.Txn, name string, entMeta *structs.EnterpriseMeta) (memdb.ResultIterator, error) {
+	return tx.Get(gatewayServicesTableName, "gateway", structs.NewServiceID(name, entMeta))
 }
 
-func (s *Store) serviceTerminatingGatewayNodes(tx *memdb.Txn, service string, entMeta *structs.EnterpriseMeta) (structs.ServiceNodes, <-chan struct{}, error) {
+// TODO(ingress): How to handle index rolling back when a config entry is
+// deleted that references a service?
+// We might need something like the service_last_extinction index?
+func (s *Store) serviceGatewayNodes(tx *memdb.Txn, service string, kind structs.ServiceKind, entMeta *structs.EnterpriseMeta) (uint64, structs.ServiceNodes, []<-chan struct{}, error) {
 	// Look up gateway name associated with the service
-	gw, err := s.serviceTerminatingGateway(tx, service, entMeta)
+	gws, err := s.serviceGateways(tx, service, entMeta)
```

Comment:
There's a bug here that I addressed in this commit:
https://github.com/hashicorp/consul/pull/7631/commits/8238acc0aeb1d7883c311af0f6ebf4d5dc600f00

A blocking query for a service that isn't already associated with a gateway will not wake up when a config entry is added later, because the `wathChans` below are only non-nil when `gws` isn't empty.

Let me know if you want to handle that here, or save it for the other PR.

Reply:
I think we can wait on pulling that in here at least until your PR is merged to minimize the rebase/merge overhead.

Reply:
I think we should merge yours first, since mine is a branch off of yours. 

I can take care of fixing that up when I rebase against master.